File: sdks/js_sdk/src/secrets/backend.ts
import type { Scope, SecretKey, SecretValue } from './types.js'

export interface SecretsBackend {
  set(scope: Scope, key: SecretKey, value: SecretValue): Promise<void>
  get(scope: Scope, key: SecretKey): Promise<string>
  list(scope?: Scope | null): Promise<Map<[string | null, string], string>>
}

export class Secrets<B extends SecretsBackend> {
  constructor(public readonly backend: B) {}
  set(scope: Scope, key: SecretKey, value: SecretValue) { return this.backend.set(scope, key, value) }
  get(scope: Scope, key: SecretKey) { return this.backend.get(scope, key) }
  list(scope?: Scope | null) { return this.backend.list(scope) }
}

File: sdks/py_sdk/sw4rm/secrets/backend.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Protocol, Tuple

from .types import Scope, SecretKey, SecretValue


class SecretsBackend(Protocol):
    def set(self, scope: Scope, key: SecretKey, value: SecretValue) -> None: ...
    def get(self, scope: Scope, key: SecretKey) -> str: ...
    def list(self, scope: Scope | None = None) -> Dict[Tuple[str | None, str], str]: ...


@dataclass
class Secrets:
    backend: SecretsBackend

    def set(self, scope: Scope, key: SecretKey, value: SecretValue) -> None:
        return self.backend.set(scope, key, value)

    def get(self, scope: Scope, key: SecretKey) -> str:
        return self.backend.get(scope, key)

    def list(self, scope: Scope | None = None) -> Dict[tuple[str | None, str], str]:
        return self.backend.list(scope)
